### PR TITLE
8253640: Make MEMFLAGS an enum class

### PR DESCRIPTION
**Pull request by stefank**
I'd like to propose that we make MEMFLAGS an enum class, to prevent them from being implicitly converted to integers.

I've been refactoring code that uses MEMFLAGS and have found that the functions often take integers and/or use default values. Because of this the compiler won't always tell you if you mess up the order. If we make the MEMFLAGS more type safe, the compiler will complain immediately if you send in the arguments in the wrong order.

Changes in the patch:

1) G1 is the only code that uses MemoryType instead of the MEMFLAGS typedef. I changed those to make the code base uniform w.r.t. this. 
2) Added some code generation so that we still can use mtGC, and don't have to write MEMFLAGS::mtGC.
3) Some stricter checks around the values stored in MEMFLAGS variables. There was a masking of 0xFF that I removed in favor of an assert. Need to run some more extensive testing see if there was a need for this masking.
4) Removed unused by_index

Notes: 
1) I didn't specify the underlying type of the enum, but it's clear for its usage that code will start to break if the value range is larger than 8 bits. Maybe something to clean up in a separate patch?
2) The code sometimes talks about "flags" and sometimes about "memory type", it's quite clear that the name changed, but the comments and names didn't get cleaned up accordingly.

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Issue
 * [JDK-8253640](https://bugs.openjdk.java.net/browse/JDK-8253640): Make MEMFLAGS an enum class

### Reviewers
 * [Thomas Stuefe](https://openjdk.java.net/census#stuefe) (@tstuefe - **Reviewer**)
 * [Thomas Schatzl](https://openjdk.java.net/census#tschatzl) (@tschatzl - **Reviewer**)

### Download
`$ git fetch https://git.openjdk.java.net/jdk pull/378/head:pull/378`
`$ git checkout pull/378`

### CODE REVIEW

File: src/hotspot/share/services/nmtCommon.hpp [hunk 1]
```diff
@@ -47,11 +47,21 @@ const int NMT_TrackingStackDepth = 4;
 // A few common utilities for native memory tracking
 class NMTUtil : AllStatic {
  public:
+  // Check if index is a valid MEMFLAGS enum value
+  static inline bool flag_index_is_valid(int index) {
+    return index >= 0 && index < mt_number_of_types;
+  }
+
+  // Check if flag value is a valid MEMFLAGS enum value
```
Comment (tstuefe) [inline src/hotspot/share/services/nmtCommon.hpp:55]: Small nit, can you please clarify that this includes "mtNone"?

File: src/hotspot/share/memory/allocation.hpp [hunk 2]
```diff
@@ -151,14 +151,20 @@ class AllocatedObj {
 /*
  * Memory types
  */
-enum MemoryType {
+enum class MEMFLAGS {
```
Comment (tstuefe) [inline src/hotspot/share/memory/allocation.hpp:154]: Remark (possible for a later change): As you wrote, I think using uint8 would be okay and safer here.

**Review by tstuefe — APPROVED** [review]
Hi Stefan,

I was just looking this morning at this enum and wished it were a class. This will make reworking the reservation APIs less error prone.

Found some small nits, but leave it up to you if you change anything.

(For some reason I always believed that using enum class would prevent us from using sentinel values but for the life of me I cannot remember why I thought this...)

Thanks, Thomas

File: src/hotspot/share/services/nmtCommon.hpp (hunk 1 above)
Comment (stefank) [reply src/hotspot/share/services/nmtCommon.hpp:55]: Good point!

**Review by tstuefe — APPROVED** [review]
lgtm

**Comment by stefank** [comment]
Thanks!

**Review by tschatzl — APPROVED** [review]
Lgtm.

**Comment by stefank** [comment]
Thanks, Thomas!

**Pull request closed without merging**